frontend/canonical.php

```diff
@@ -0,0 +1,261 @@
+			wp_safe_redirect( $redirect_url, 301, POLYLANG );
+			exit;
+		}
+	}
+
+	/**
+	 * Returns the term_id of the requested term.
+	 *
+	 * @since 2.9
+	 *
+	 * @param WP_Tax_Query $tax_query An instance of WP_Tax_Query.
+	 * @return int
+	 */
+	protected function get_queried_term_id( $tax_query ) {
+		$queried_terms = $tax_query->queried_terms;
+		$taxonomy = $this->get_queried_taxonomy( $tax_query );
+
+		if ( ! is_array( $queried_terms[ $taxonomy ]['terms'] ) ) {
+			return 0;
+		}
+		$field = $queried_terms[ $taxonomy ]['field'];
+		$term  = reset( $queried_terms[ $taxonomy ]['terms'] );
+		$lang  = isset( $queried_terms['language']['terms'] ) ? reset( $queried_terms['language']['terms'] ) : '';
+
+		// We can get a term_id when requesting a plain permalink, eg /?cat=1.
+		if ( 'term_id' === $field ) {
+			return $term;
+		}
+
+		// We get a slug when requesting a pretty permalink. Let's query all corresponding terms.
+		$args = array(
+			'lang'       => '',
+			'taxonomy'   => $taxonomy,
+			$field       => $term,
+			'hide_empty' => false,
+			'fields'     => 'ids',
+		);
+		$term_ids = get_terms( $args );
+
+		if ( ! is_array( $term_ids ) || empty( $term_ids ) ) {
+			return 0;
+		}
+
+		$term_ids = array_filter( $term_ids, 'is_numeric' );
+
+		$filtered_terms_by_lang = array_filter(
+			$term_ids,
+			function ( $term_id ) use ( $lang ) {
+				$term_lang = $this->model->term->get_language( (int) $term_id );
+
+				return ! empty( $term_lang ) && $term_lang->slug === $lang;
+			}
+		);
+
+		$tr_term = (int) reset( $filtered_terms_by_lang );
+
+		if ( ! empty( $tr_term ) ) {
+			// The queried term exists in the desired language.
+			return $tr_term;
+		}
+
+		// The queried term doesn't exist in the desired language, let's return the first one retrieved.
+		return (int) reset( $term_ids );
+	}
+
+	/**
+	 * Find the taxonomy being queried.
+	 *
+	 * @since 2.9
+	 *
+	 * @param WP_Tax_Query $tax_query An instance of WP_Tax_Query.
+	 * @return string A taxonomy slug
+	 */
+	protected function get_queried_taxonomy( $tax_query ) {
+		$queried_terms = $tax_query->queried_terms;
+		unset( $queried_terms['language'] );
+
+		return (string) key( $queried_terms );
+	}
+
+	/**
+	 * Evaluates the canonical redirect url through the deidcated WP function.
+	 *
+	 * @since 3.3
+	 *
+	 * @param string       $url      Requested url.
+	 * @param PLL_Language $language Language of the queried object.
+	 * @return string
+	 */
+	protected function redirect_canonical( $url, $language ) {
+		global $wp_query;
+
+		$this->curlang = $language; // Hack to filter the `page_for_posts` option in the correct language.
+
+		$backup_wp_query = $wp_query;
+
+		if ( isset( $wp_query->tax_query ) ) {
+			unset( $wp_query->tax_query->queried_terms['language'] );
+			unset( $wp_query->query['lang'] );
+		}
+
+		$redirect_url = redirect_canonical( $url, false );
```

Comment:
I wonder if unsetting the language from the query could lead to problem in `redirect_canonical()` for shared slugs?